### ISSUE TITLE
<Leader>hs error with pwsh on W10

### ISSUE THREAD

**Issue by Acelya-9028** [airblade/vim-gitgutter#751]
Hello,

I use Neovim and pwsh on W10 and when I use `<Leader>hs`, `<Leader>hu` or `<Leader>hp` gitgutter crash.

![gitgutter_error](https://user-images.githubusercontent.com/45818681/101698784-63d11500-3a7a-11eb-8b86-43007107af2b.png)

Here is my shell config:

```
set shell=pwsh shellpipe=\| shellxquote=
set shellcmdflag=-NoLogo\ -NoProfile\ -ExecutionPolicy\ RemoteSigned\ -Command
set shellredir=\|\ Out-File\ -Encoding\ UTF8
```

I know that it's a problem with powershell when a temp file is used instead pipe because I have a similar probkem with [fugitive](https://github.com/tpope/vim-fugitive/issues/1644).

fugitive works with `set noshelltemp`  but gitgutter not :s.
gitgutter works with cmd.exe, but not with powershell or pwsh, with or without `set noshelltemp`, error is the same...

```
NVIM v0.4.4
Build type: RelWithDebInfo
LuaJIT 2.0.5
Compilation: C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x64/cl.exe /DWIN32 /D_WINDOWS /W3 /MD /Zi /O2 /Ob1 /DNDEBUG -DMIN_LOG_LEVEL=3 /W3 -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_DEPRECATE -DWIN32 -D_WIN32_WINNT=0x0600 -DINCLUDE_GENERATED_DECLARATIONS -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -IC:/projects/neovim/build/config -IC:/projects/neovim/src -IC:/projects/nvim-deps/usr/include -IC:/projects/neovim/build/src/nvim/auto -IC:/projects/neovim/build/include
Compilé par appveyor@APPVYR-WIN

Features: -acl +iconv +tui
See ":help feature-compile"

         fichier vimrc système : "$VIM\sysinit.vim"
               $VIM par défaut : "C:/Program Files/nvim/share/nvim"

Run :checkhealth for more info
```

```
Nom du système d’exploitation:      Microsoft Windows 10 Entreprise LTSC
Version du système:                        10.0.17763 N/A version 17763
```

```
gitgutter: commit dfe55e2b924b86c654b63edb9bedc42aa4e08048
```

Thank you very much for your help

**Comment by airblade** [comment]
I'll try to help but I haven't used Windows for 15+ years so I really don't know how it works ;)

You could try switching on logging with `:let g:gitgutter_log=1` (the log file will be in gitgutter's own directory) and seeing if anything helpful is logged when you try a hunk operation.

You could also try `:let g:gitgutter_async=0` as that executes external commands slightly differently.

**Comment by Acelya-9028** [comment]
Ahah, I understand you, it annoys me a bit actually. Thank you for the logging switch, it's very useful.

Here is a log (with shell set to  `pwsh`):

```
--- Sign Column Part (Nice)

0.361677 function <SNR>22_on_exit_nvim[2]..1[5]..gitgutter#process_buffer[19]..gitgutter#diff#run_diff[84]..gitgutter#async#execute[1]:
  0.361677 [async] cd "C:\home\projects\vim-gitgutter\autoload\gitgutter" && (git  --no-pager show :autoload/gitgutter/diff.vim > C:\Users\K2-18b\AppData\Local\Temp\nvimS0d2Xv\1.12.2.vim && git  --no-pager -c "diff.autorefreshindex=0" -c "diff.noprefix=false" -c "core.safecrlf=false" diff --no-ext-diff --no-color -U0  -- C:\Users\K2-18b\AppData\Local\Temp\nvimS0d2Xv\1.12.2.vim C:\Users\K2-18b\AppData\Local\Temp\nvimS0d2Xv\2.12.2.vim || exit 0)

  0.418631 function <SNR>22_on_exit_nvim[2]..gitgutter#diff#handler[1]:
  0.418631 diff --git a/C:/Users/K2-18b/AppData/Local/Temp/nvimS0d2Xv/1.12.1.vim b/C:/Users/K2-18b/AppData/Local/Temp/nvimS0d2Xv/2.12.1.vim
index d88f8f0..6beb019 100644
--- a/C:/Users/K2-18b/AppData/Local/Temp/nvimS0d2Xv/1.12.1.vim
+++ b/C:/Users/K2-18b/AppData/Local/Temp/nvimS0d2Xv/2.12.1.vim
@@ -123,0 +124 @@ function! gitgutter#diff#run_diff(bufnr, from, preserve_full_diff) abort
+    " let cmd .= g:gitgutter_git_executable.' '.g:gitgutter_git_args.' --no-pager show '.index_name.' > '.from_file.' && (((Get-Content '.from_file.') -join "`n") + "`n" | Set-Content -NoNewline '.from_file.') && '

--- <leader>hs Part (Not Nice)

411.326232 function gitgutter#hunk#preview[3]..<SNR>106_hunk_op[36]..gitgutter#diff#run_diff[91]..gitgutter#utility#system[1]:
411.326232 cd "C:\home\projects\vim-gitgutter\autoload\gitgutter" && (git  --no-pager show :autoload/gitgutter/diff.vim > C:\Users\K2-18b\AppData\Local\Temp\nvimS0d2Xv\1.12.6.vim && git  --no-pager -c "diff.autorefreshindex=0" -c "diff.noprefix=false" -c "core.safecrlf=false" diff --no-ext-diff --no-color -U0  -- C:\Users\K2-18b\AppData\Local\Temp\nvimS0d2Xv\1.12.6.vim C:\Users\K2-18b\AppData\Local\Temp\nvimS0d2Xv\2.12.6.vim || exit 0)

411.865159 function gitgutter#hunk#preview[3]..<SNR>106_hunk_op[36]..gitgutter#diff#run_diff[94]:
411.865159 ^[[91mexit: ^[[91mThe term 'exit' is not recognized as a name of a cmdlet, function, script file, or executable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.^[[0m
diff --git a/C:/Users/K2-18b/AppData/Local/Temp/nvimTYPwyu/1.12.6.vim b/C:/Users/K2-18b/AppData/Local/Temp/nvimTYPwyu/2.12.6.vim
index 2796f7e..6beb019 100644
--- a/C:/Users/K2-18b/AppData/Local/Temp/nvim70hejK/1.12.6.vim
+++ b/C:/Users/K2-18b/AppData/Local/Temp/nvim70hejK/2.12.6.vim
@@ -1,417 +1,418 @@
-scriptencoding utf8
-
-let s:nomodeline = (v:version > 703 || (v:version == 703 && has('patch442'))) ? 
-----------------------------------------------
[[BIG CUT BECAUSE IT'S THE WHOLE FILE * 2]]
-----------------------------------------------
+function! s:save_last_seen_change(bufnr) abort
+  call gitgutter#utility#setbufvar(a:bufnr, 'tick', getbufvar(a:bufnr, 'changedtick'))
+endfunction
```
The column sign is done asynchronously with `gitgutter#async#execute` who use `jobstart()` with a list (because neovim), and according to the doc, use `cmd` and it's good.

But for the maps, `gitgutter#utility#system` is used, so `system()` so use `pwsh`.

And here, there are two problems:

```
exit: The term 'exit' is not recognized as a name of a cmdlet, function, script file, or executable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
```

It pass the `if v:shell_error` in `gitgutter#diff#run_diff`, because above I presume.

Seriously, I looked for I didn't find a solution... I don't know how return properly an exit code in Powershell without use `exit`...

Then the broken diff. The diff if broken because `git  --no-pager show :autoload/gitgutter/diff.vim > C:\Users\K2-18b\AppData\Local\Temp\nvimY4zY3V\0.12.4.vim` with `cmd` use `\n` but `\r\n` with `pwsh` cf [PowerShell/issues/2145](https://github.com/PowerShell/PowerShell/issues/2145#issuecomment-243873278).

I think it's fixable by forcing the use of `cmd` on Windows, or by replacing `\r\n` by `\n` before doing the diff like below, but it's dirty and it works only with Powershell, not `cmd`.

```
let cmd .= g:gitgutter_git_executable.' '.g:gitgutter_git_args.' --no-pager show '.index_name.' > '.from_file.' && (((Get-Content '.from_file.') -join "`n") + "`n" | Set-Content -NoNewline '.from_file.') && '
```

Have you an idea?

**Comment by airblade** [comment]
Thanks for the detailed analysis!

If you do `set shell="cmd.exe"` just before `<leader>hs`, does it all work?

**Comment by Acelya-9028** [comment]
Hum, `:set shell="cmd.exe"` only, no because the result is broken but `set shell=cmd.exe shellcmdflag=/s\ /c shellredir=>%s\ 2>&1 shellpipe=>%s\ 2>&1 shellxquote="` so a full reset, yes.

Is it possible to do the same things as `unix` for Powershell in `s:use_known_shell()` and `s:restore_shell()` ? I can pr that.

**Comment by airblade** [comment]
> Is it possible to do the same things as unix for Powershell in `s:use_known_shell()` and `s:restore_shell()`?

That's what I was planning to do if you could get it working with cmd.exe :)

**Comment by airblade** [comment]
Fixed by #752.
